sequencer/src/bin/verify-headers.rs

```diff
@@ -0,0 +1,213 @@
+            parent.timestamp,
+            header.timestamp
+        );
+        ok = false;
+    }
+    if !opt.no_l1_heads && header.l1_head < parent.l1_head {
+        tracing::error!(
+            "header {height} has decreasing L1 head: {} -> {}",
+            parent.l1_head,
+            header.l1_head
+        );
+        ok = false;
+    }
+    if !opt.no_l1_finalized && header.l1_finalized < parent.l1_finalized {
+        tracing::error!(
+            "header {height} has decreasing L1 finalized: {:?} -> {:?}",
+            parent.l1_finalized,
+            header.l1_finalized
+        );
+        ok = false;
+
+        if let (Some(l1), Some(l1_finalized)) = (l1, &header.l1_finalized) {
+            let l1_block = get_l1_block(l1, l1_finalized.number).await;
+            if *l1_finalized != l1_block {
+                tracing::error!(
+                    "header {height} has wrong L1 finalized info: {l1_finalized:?}, expected {l1_block:?}"
+                );
+                ok = false;
+            }
+        }
+    }
+
+    (header, ok)
+}
+
+async fn get_header(seq: &SequencerClient, height: usize) -> Header {
+    loop {
+        match seq
+            .get(&format!("availability/header/{height}"))
+            .send()
+            .await
+        {
+            Ok(header) => break header,
+            Err(err) => {
+                tracing::warn!("error fetching header {height}: {err}");
+
+                // Back off a bit and then retry.
+                sleep(Duration::from_millis(100)).await;
+            }
+        }
+    }
+}
+
+async fn get_l1_block(l1: &Provider<Http>, height: u64) -> L1BlockInfo {
+    loop {
+        let block = match l1.get_block(height).await {
+            Ok(Some(block)) => block,
+            Ok(None) => {
+                tracing::warn!("L1 block {height} not yet available");
+                sleep(Duration::from_secs(1)).await;
+                continue;
+            }
+            Err(err) => {
+                tracing::warn!("error fetching L1 block {height}: {err}");
+                sleep(Duration::from_millis(100)).await;
+                continue;
+            }
+        };
+
+        let Some(hash) = block.hash else {
+            tracing::warn!("L1 block {height} has no hash, might not be finalized yet");
+            sleep(Duration::from_secs(1)).await;
+            continue;
+        };
+
+        return L1BlockInfo {
+            number: height,
+            hash,
+            timestamp: block.timestamp,
+        };
+    }
+}
+
+#[async_std::main]
+async fn main() {
+    setup_logging();
+    setup_backtrace();
+
+    let opt = Arc::new(Options::parse());
+    let seq = Arc::new(SequencerClient::new(opt.url.clone()));
+    seq.connect(None).await;
+
+    let block_height: usize = seq.get("status/latest_block_height").send().await.unwrap();
+    let from = opt.from.unwrap_or(0);
+    let to = max(opt.to.unwrap_or(block_height), from + 1);
+
+    let ok = join_all(
```

Comment:
I think it would be nice if this logged something about the progress on INFO level or had something like a progress bar. Currently there's no way to distinguish the long running working output from the one where the URL is wrong (both log nothing).

Reply:
I did something slightly different ([919f753](https://github.com/EspressoSystems/espresso-sequencer/pull/869/commits/919f7534f69516ac5a50cfd431c1316c348d56e6)). I got rid of the `seq.connect()` call, which blocked until a successful connection, so it now panics if the URL is wrong. After that, it will print out an INFO level log telling you the block range it is going to process, and if anything goes wrong it logs something at WARN level. The per-header logs remain at DEBUG level so it's not too noisy and it's easy to see if there are errors.